[JSC] Avoid checking if any function parameter is captured unless necessary

shvaikalesh opened the pull request:
#### 0ba44cd43a98c967e1a629f0134447fa8ff88a34
<pre>
[JSC] Avoid checking if any function parameter is captured unless necessary
<a href="https://bugs.webkit.org/show_bug.cgi?id=266015">https://bugs.webkit.org/show_bug.cgi?id=266015</a>
&lt;<a href="https://rdar.apple.com/problem/119337256">rdar://problem/119337256</a>&gt;

Reviewed by Mark Lam.

With this change, `capturesAnyParameterByName` is computed only if it&apos;s actually used,
guarded by condition that includes `isSimpleParameterList`, which enables us to assert
isBindingNode().

Also, the loop is now terminated right after first captured parameter is found.

No new tests, no behavior change.

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):

Canonical link: <a href="https://commits.webkit.org/271719@main">https://commits.webkit.org/271719@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/467b06e3f7bbb704ec8ad64bae68d65c5afcae85

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29426 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30748 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31974 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26689 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10229 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5384 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26686 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29699 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6723 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25146 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5785 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26225 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33314 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/25329 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26844 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26638 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32132 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/29651 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5872 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4053 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29909 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7586 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36001 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6399 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7758 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3784 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6396 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (approved):
r=me if EWS bots are green.

MenloDorian commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -576,8 +560,21 @@ BytecodeGenerator::BytecodeGenerator(VM& vm, FunctionNode* functionNode, Unlinke
         // activation.
+        bool capturesAnyParameterByName = false;
+        if (functionNode->hasCapturedVariables()) {
+            for (size_t i = 0; i < parameters.size(); ++i) {
```
We used to fetch `parameters` from `functionNode->parameters()`.  Why is it ok to use this version of parameters rather than th one from the `functionNode`?

shvaikalesh replied:
It's totally ok since the `parameters` is defined as

```cpp
FunctionParameters& parameters = *functionNode->parameters();
```

and it's not being mutated.

MenloDorian commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -576,8 +560,21 @@ BytecodeGenerator::BytecodeGenerator(VM& vm, FunctionNode* functionNode, Unlinke
+        if (functionNode->hasCapturedVariables()) {
+            for (size_t i = 0; i < parameters.size(); ++i) {
+                auto pattern = parameters.at(i).first;
+                ASSERT(pattern->isBindingNode());
```
Why does `isSimpleParameterList` guarantee that `pattern->isBindingNode()` is true?

shvaikalesh replied:
It's guaranteed per implementation of `FunctionParameters::append()`:

```cpp
            bool isSimpleParameter = !hasDefaultParameterValue && pattern->isBindingNode();
            m_isSimpleParameterList &= isSimpleParameter;
```

Commit:
[JSC] Avoid checking if any function parameter is captured unless necessary
https://bugs.webkit.org/show_bug.cgi?id=266015
<rdar://problem/119337256>

Reviewed by Mark Lam.

With this change, `capturesAnyParameterByName` is computed only if it's actually used,
guarded by condition that includes `isSimpleParameterList`, which enables us to assert
isBindingNode().

Also, the loop is now terminated right after first captured parameter is found.

No new tests, no behavior change.

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):

Canonical link: https://commits.webkit.org/271719@main
